### ISSUE TITLE
`too-many-function-args` false positive when decorator uses `ParamSpec` and `Concatenate` to change function arguments

### ISSUE THREAD

**Issue by DetachHead** [PyCQA/pylint#8175]
### Bug description

```py
from typing import Callable, Concatenate, ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")


def foo(fn: Callable[P, R]) -> Callable[Concatenate[str, P], R]:
    def wrapped(value: str, *args: P.args, **kwargs: P.kwargs) -> R:
        print(value)
        return fn(*args, **kwargs)

    return wrapped


@foo
def bar():
    ...


# Too many positional arguments for function call (E1121:too-many-function-args)
bar("hi")  # works at runtime
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:21:0: E1121: Too many positional arguments for function call (too-many-function-args)
```

### Expected behavior

no error

### Pylint version

```shell
pylint 2.16.0b1
astroid 2.14.0-dev0
Python 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]
```

**Comment by DanielNoord** [comment]
This is just too dynamic for `pylint` to statically infer. I wonder if we should even keep this issue open as I don't see us fixing it at some point..

**Comment by KotlinIsland** [comment]
The implementation transform is quite dynamic, but it is comprehensively described by the type signature, therefore it is entirely statically verifiable.

**Comment by DanielNoord** [comment]
Pylint doesn't use the type annotations. There are multiple issues about this already, but currently that would be a major rewrite of the tool.

**Comment by Pierre-Sassoulas** [comment]
This would need #4813 to work, closing as "not feasible until then".
